pyop2/arg.py

```diff
@@ -0,0 +1,194 @@
+        # if configuration["type_check"] and not (self._is_global or map is None):
+        #     for j, m in enumerate(map):
+        #         if m.iterset.total_size > 0 and len(m.values_with_halo) == 0:
+        #             raise MapValueError("%s is not initialized." % map)
+        #         if self._is_mat and m.toset != data.sparsity.dsets[j].set:
+        #             raise MapValueError(
+        #                 "To set of %s doesn't match the set of %s." % (map, data))
+        #     if self._is_dat and map.toset != data.dataset.set:
+        #         raise MapValueError(
+        #             "To set of %s doesn't match the set of %s." % (map, data))
+
+    @property
+    def _wrapper_cache_key_(self):
+        return self._key
+
+    @property
+    def _key(self):
+        return type(self), self.access
+
+    def __eq__(self, other):
+        r""":class:`Arg`\s compare equal of they are defined on the same data,
+        use the same :class:`Map` with the same index and the same access
+        descriptor."""
+        return self._key == other._key
+
+    def __ne__(self, other):
+        r""":class:`Arg`\s compare equal of they are defined on the same data,
+        use the same :class:`Map` with the same index and the same access
+        descriptor."""
+        return not self.__eq__(other)
+
+    def __str__(self):
+        return "OP2 Arg: dat %s, access %s" % \
+            (self.data_class, self.access)
+
+    def __repr__(self):
+        return "Arg(%r, %r)" % \
+            (type(self), self.access)
+
+    @cached_property
+    def name(self):
+        """The generated argument name."""
+        return "arg%d" % self.position
+
+    @cached_property
+    def ctype(self):
+        """String representing the C type of the data in this ``Arg``."""
+        return self.data.ctype
+
+    @cached_property
+    def dtype(self):
+        """Numpy datatype of this Arg"""
+        return self._dtype
+
+
+class DatArg(Arg):
+
+    def __init__(self, access, dtype, shape, is_direct, map_shape=None, map_offset=None):
+        super().__init__(access, dtype)
+        self.shape = shape
+        self.is_direct = is_direct
+        self.map_shape = map_shape
+        self.map_offset = map_offset
+
+    @property
+    def is_indirect(self):
+        return not self.is_direct
+
+
+class DatViewArg(DatArg):
+
+    def __init__(self, access, dtype, shape, is_direct, index, map_shape=None, map_offset=None):
+        super().__init__(access, dtype, shape)
+        self.index = index
+
+
+class GlobalArg(Arg):
+
+    def __init__(self, access, dtype, dim):
+        super().__init__(access, dtype)
+
+        self.dim = dim
+
+    @property
+    def is_reduction(self):
+        return self._access in {INC, MIN, MAX}
+
+
+class MatArg(Arg):
+
+    def __init__(self, access, dtype, unroll_map, dims, lgmaps, map_shapes, map_offsets):
+        super().__init__(access, dtype)
+        if lgmaps is not None:
+            lgmaps = as_tuple(lgmaps)
```

Comment:
These are heavy data and so shouldn't live here I think.

Reply:
Didn't know that. Thanks.